lib/fizzy/leb128.hpp

```diff
@@ -46,13 +49,31 @@ std::pair<T, const uint8_t*> leb128s_decode(const uint8_t* input, const uint8_t*
         result |= static_cast<T_unsigned>((static_cast<T_unsigned>(*input) & 0x7F) << result_shift);
         if ((*input & 0x80) == 0)
         {
-            // sign extend
-            if ((*input & 0x40) != 0 && (result_shift + 7 < sizeof(T_unsigned) * 8))
+            constexpr auto all_ones = std::numeric_limits<T_unsigned>::max();
+
+            if (result_shift + 7 < sizeof(T_unsigned) * 8)
             {
-                constexpr auto all_ones = std::numeric_limits<T_unsigned>::max();
-                const auto mask = static_cast<T_unsigned>(all_ones << (result_shift + 7));
-                result |= mask;
+                // non-terminal byte of the encoding, extend the sign bit, if needed
+                if ((*input & 0x40) != 0)
+                {
+                    const auto mask = static_cast<T_unsigned>(all_ones << (result_shift + 7));
+                    result |= mask;
+                }
             }
+            else
+            {
+                // terminal byte of the encoding, need to check unused bits
+                const auto unused_bits_mask = ~static_cast<uint8_t>(all_ones >> result_shift);
+                // unused bits must be set if and only if sign bit is set
+                const auto unused_bits_expected =
+                    static_cast<T>(result) < 0 ? (unused_bits_mask & 0x7F) : 0;
+                if ((*input & unused_bits_mask) != unused_bits_expected)
```

Comment:
You can also use shift to right here, but arithmetic one this time.
```suggestion
                if (*input == static_cast<uint8_t>(static_cast<T>(result) >> result_shift))
```

Reply:
But right-shifting negative values is implementation-dependent, that's why I avoided it here.

Reply:
I can add a static_assert to check the compiler.

Reply:
You are right. I learned about this only recently and I keep forgetting. Anyway, UBSan seems to be happy.